src/transformStateWithClones.js

```diff
@@ -7,7 +7,46 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const resultArray = [];
+  let removePropertiesClone = {};
+
+  for (let i = 0; i < actions.length; i++) {
+    switch (actions[i].type) {
+      case 'addProperties':
+        switch (resultArray.length) {
+          case 0:
+            resultArray.push(Object.assign({}, state, actions[i].extraData));
+            break;
+          default:
+            resultArray.push(Object.assign({
+            }, resultArray[i - 1], actions[i].extraData));
+        }
+        break;
+      case 'removeProperties':
+        switch (resultArray.length) {
+          case 0:
+            removePropertiesClone = Object.assign({}, state);
+
+            for (const keyToRemove of actions[i].keysToRemove) {
+              delete removePropertiesClone[keyToRemove];
+            }
+            resultArray.push(removePropertiesClone);
+            break;
+          default:
+            removePropertiesClone = Object.assign({}, resultArray[i - 1]);
+
+            for (const keyToRemove of actions[i].keysToRemove) {
```

Comment:
why do you need such complicated structure?

Reply:
you should iterate through actions and then switch through `action.type` and describe logic into cases